Ical.Net/DataTypes/CalDateTime.cs

```diff
@@ -414,53 +414,91 @@ public DateTime Value
+        // of discontinuities in the time scale, such as the change from standard time to daylight
+        // time and back, the computation of the exact duration requires the subtraction or
+        // addition of the change of duration of the discontinuity.Leap seconds MUST NOT be
+        // considered when computing an exact duration.When computing an exact duration, the
+        // greatest order time components MUST be added first, that is, the number of days MUST be
+        // added first, followed by the number of hours, number of minutes, and number of seconds.
+
+        (TimeSpan? nominalPart, TimeSpan? exactPart) dt;
+        if (TzId is null)
+            dt = (d.ToTimeSpan(), null);
+        else
+            dt = (d.HasDate ? d.DateAsTimeSpan : null, d.HasTime ? d.TimeAsTimeSpan : null);
+
+        var newDateTime = this;
+        if (dt.nominalPart is not null)
+            newDateTime = new CalDateTime(newDateTime.Value.Add(dt.nominalPart.Value), TzId, HasTime);
 
-        var copy = Copy<CalDateTime>();
-        copy._dateOnly = DateOnly.FromDateTime(newDateTime);
-        copy._timeOnly = HasTime ? TruncateTimeToSeconds(newDateTime) : null;
-        copy._tzId = TzId is null ? null : UtcTzId;
+        if (dt.exactPart is not null)
+            newDateTime = new CalDateTime(newDateTime.AsUtc.Add(dt.exactPart.Value), UtcTzId, HasTime);
+        
+        if (TzId is not null)
+            // Convert to the original timezone even if already set to ensure we're not in a non-existing time.
+            newDateTime = newDateTime.ToTimeZone(TzId);
 
-        return TzId is null ? copy : copy.ToTimeZone(TzId);
+        AssociateWith(this);
+
+        return newDateTime;
     }
 
-    /// <summary>Returns a new <see cref="TimeSpan" /> from subtracting the specified <see cref="IDateTime"/> from to the value of this instance.</summary>
-    /// <param name="dt"></param>
-    public TimeSpan Subtract(IDateTime dt) => AsUtc - dt.AsUtc;
+    /// <inheritdoc/>
+    public TimeSpan SubtractExact(IDateTime dt) => AsUtc - dt.AsUtc;
 
-    /// <summary>Returns a new <see cref="IDateTime"/> by subtracting the specified <see cref="TimeSpan" /> from the value of this instance.</summary>
-    /// <param name="ts">An interval.</param>
-    /// <returns>An object whose value is the difference of the date and time represented by this instance and the time interval represented by <paramref name="ts" />.</returns>
-    /// <exception cref="InvalidOperationException">
-    /// Thrown when attempting to subtracting a time span from a date-only instance, 
-    /// and the time span is not a multiple of full days.
-    /// </exception>
-    public IDateTime Subtract(TimeSpan ts) => Add(-ts);
+    /// <inheritdoc/>
+    public Duration Subtract(IDateTime dt)
+    {
+        if (this.TzId is not null)
+            return SubtractExact(dt).ToDurationExact();
+
+        if (dt.HasTime != HasTime)
```

Comment:
good

Reply:
Yeah, I also tried throwing if `IsFloating != dt.IsFloating`, but then all kinds of tests fail. Obviously there are quite some cases where `DTSTART` has a tzid set but `DTEND` hasn't, which isn't allowed by the RFC though. We should double-check whether we still want to support this case (probably yes, if its used in the wild). If so, we should probably deal with the case in a different place, probably during deserialization, and be more strict here, because the current implementation is somewhat unclear regarding that case.